src/fast_layers.jl

```diff
@@ -72,6 +72,87 @@ struct FastDense{F,F2,C} <: FastLayer
   end
 end
 
+# To work with scalars(x::Number)
+(f::FastDense)(x::Number,p) = ((f.bias == true) ? (f.σ.(reshape(p[1:(f.out*f.in)],f.out,f.in)*x .+ p[(f.out*f.in+1):end])) : (f.σ.(reshape(p[1:(f.out*f.in)],f.out,f.in)*x)))
+
+ZygoteRules.@adjoint function (f::FastDense)(x::Number,p)
+  if typeof(f.cache) <: Nothing
+    if !isgpu(p)
+      W = @view(p[reshape(1:(f.out*f.in),f.out,f.in)])
+    else
+      W = reshape(@view(p[1:(f.out*f.in)]),f.out,f.in)
+    end
+    if f.bias == true
+      b = p[(f.out*f.in + 1):end]
+      r = W*x .+ b
+      ifgpufree(b)
+    else
+      r = W*x
+    end
+    y = f.σ.(r)
+  else
+    if !isgpu(p)
+      f.cache.W .= @view(p[reshape(1:(f.out*f.in),f.out,f.in)])
+    else
+      f.cache.W .= reshape(@view(p[1:(f.out*f.in)]),f.out,f.in)
+    end
+    mul!(@view(f.cache.r[:,1]), f.cache.W, x)
+    if f.bias == true
+      # @view(f.cache.r[:,1]) .+= @view(p[(f.out*f.in + 1):end])
+      b = @view(p[(f.out*f.in + 1):end])
+      @view(f.cache.r[:,1]) .+= b
+    end
+    f.cache.yvec .= f.σ.(@view(f.cache.r[:,1]))
+  end
+  function FastDense_adjoint(ȳ)
+    if typeof(f.cache) <: Nothing
+      if typeof(f.σ) <: typeof(NNlib.tanh_fast)
+        zbar = ȳ .* (1 .- y.^2)
+      elseif typeof(f.σ) <: typeof(identity)
+        zbar = ȳ
+      else
+        zbar = ȳ .* ForwardDiff.derivative.(f.σ,r)
+      end
+      Wbar = zbar * x'
+      bbar = zbar
+      xbar = W' * zbar
+      pbar = if f.bias == true
+          tmp = typeof(bbar) <: AbstractVector ?
+                           vec(vcat(vec(Wbar),bbar)) :
+                           vec(vcat(vec(Wbar),sum(bbar,dims=2)))
+          ifgpufree(bbar)
+          tmp
+      else
+          vec(Wbar)
+      end
+      ifgpufree(Wbar)
+      ifgpufree(r)
+      nothing,xbar,pbar
+    else
+      if typeof(f.σ) <: typeof(NNlib.tanh_fast)
+        @view(f.cache.zbar[:,1]) .= ȳ .* (1 .- (f.cache.yvec).^2)
+      elseif typeof(f.σ) <: typeof(identity)
+        @view(f.cache.zbar[:,1]) .= ȳ
+      else
+        @view(f.cache.zbar[:,1]) .= ȳ .* ForwardDiff.derivative.(f.σ, @view(f.cache.r[:,1]))
+      end
+      mul!(f.cache.Wbar, @view(f.cache.zbar[:,1]), x')
+      mul!(@view(f.cache.xbar[:,1]), f.cache.W', @view(f.cache.zbar[:,1]))
+      f.cache.pbar .= if f.bias == true
+        vec(vcat(vec(f.cache.Wbar),@view(f.cache.zbar[:,1])))# bbar = zbar
+      else
+        vec(f.cache.Wbar)
+      end
+      nothing,@view(f.cache.xbar[:,1]),f.cache.pbar

```

Comment:
The derivative w.r.t. `x` should make sure to return a scalar.

Reply:
Whoops, I see it now in first 2 lines of the error